### PR TITLE
zombienet-sdk-tests: migrate smoke tests from v1 to SDK

### PR DESCRIPTION
**Pull request by DenzelPenzel**
## Description

#10614 

Migrate polkadot smoke tests from the legacy zombienet-v1 (zndsl format) to the new
zombienet-sdk:
- **0001-parachains-smoke-test**: Verifies parachain registration and block production
- **0002-parachains-parachains-upgrade-smoke(flaky)**: Verifies that a parachain can be registered, produce blocks
- **0003-deregister-register-validator**: Tests dynamic validator set changes via validatorManager pallet
- **0004-coretime-smoke-test(flaky)**: Validates bulk coretime assignment and parachain onboarding
- **0007-dispute-freshly-finalized**: Verifies that disputes triggered on finalized blocks within scope always end as valid
- **0005-precompile-pvf-smoke(flaky)**: Verifies that PVF precompilation behaves correctly when validators are added/removed from the validator set.
- **0006-parachains-max-tranche0**: Verifies that parachains make progress with most of the approvals being tranche0
- **0014-chunk-fetching-network-compatibility(flaky)**: Verifies that validators preserve backwards compatibility with peers speaking an older version of the /req_chunk protoco
- **0015-coretime-shared-core**: Verifies that multiple parachains can share a single core using coretime assignments
- **0019-coretime-collation-fetching-fairness**: Verifies the fairness of block production by monitoring CandidateIncluded events.

## Changes

- Add `utils.rs` with shared utilities (metric constants, genesis header/validation code fetching, sudo call builders)
- Add three new Rust test files implementing the SDK versions
- Update CI configuration to use `use-zombienet-sdk: true` for migrated tests

## Test results
- **0001-parachains-smoke-test** (30 runs no errors found)
- **0003-deregister-register-validator** (30 runs no errors found)
- **0007-dispute-freshly-finalized** (30 runs no errors found)

### CODE REVIEW

File: polkadot/zombienet-sdk-tests/tests/utils.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		"Sudo",
+		"sudo",
+		vec![value! {
+			ValidatorManager(register_validators { validators: (stash_account) })
+		}],
+	)
+}
+
+/// Creates a sudo batch call to register a parachain with trusted validation code
+#[allow(dead_code)]
```
Comment (bkchr) [inline polkadot/zombienet-sdk-tests/tests/utils.rs]: Why is this dead code?

Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/utils.rs]: I have temp marked this as dead code; I plan to remove it once the zombienet-polkadot-smoke-0005-precompile-pvf-smoke test is migrated

File: polkadot/zombienet-sdk-tests/tests/functional/parachains_pvf.rs [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+// PVF metric constants
+const MAX_PVF_PREP_TIME_SECS: f64 = 10.0;
+const MAX_PVF_EXEC_TIME_SECS: f64 = 2.0;
+
+// Metric bucket definitions
+const PVF_PREP_BUCKETS_ACCEPTABLE: &[&str] = &["0.1", "0.5", "1", "2", "3", "10"];
+const PVF_PREP_BUCKETS_UNACCEPTABLE: &[&str] = &["20", "30", "60", "120", "+Inf"];
+const PVF_EXEC_BUCKETS_ACCEPTABLE: &[&str] = &["0.1", "0.5", "1", "2"];
+const PVF_EXEC_BUCKETS_UNACCEPTABLE: &[&str] = &["4", "5", "6", "+Inf"];
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/parachains_pvf.rs:47]: These are _histograms_ and we should add the logic to handle in _zombienet-sdk_ like we do in v1 (https://github.com/paritytech/zombienet/blob/main/javascript/packages/orchestrator/src/metrics/index.ts#L60)

File: polkadot/zombienet-sdk-tests/tests/utils.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	network.detach().await;
+
+	Ok(network)
+}
+
+pub fn env_or_default(var: &str, default: &str) -> String {
+	std::env::var(var).unwrap_or_else(|_| default.to_string())
+}
+
+/// Fetches the genesis header from a parachain node
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/utils.rs:67]: We may want to move these methods to https://github.com/paritytech/zombienet-sdk/tree/main/crates/orchestrator/src/tx_helper, since could be useful in other places.

File: polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+}
+
+fn build_network_config() -> Result<NetworkConfig, anyhow::Error> {
+	let images = zombienet_sdk::environment::get_images_from_env();
+	let polkadot_image = env_or_default(INTEGRATION_IMAGE_ENV, images.polkadot.as_str());
+	let old_suffix = std::env::var("OLD_SUFFIX").unwrap_or_default();
+
+	// Old image that doesn't speak /req_chunk/2 protocol
+	let old_polkadot_image = std::env::var("POLKADOT_IMAGE")
+		.map(|img| format!("{}:master-bde0bbe5", img.split(':').next().unwrap_or(&img)))
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs]: can we make the _old tag_  (master-bde0bbe5) const

File: polkadot/zombienet-sdk-tests/tests/functional/parachains_pvf.rs (hunk 2 above)
Comment (mchristou) [reply polkadot/zombienet-sdk-tests/tests/functional/parachains_pvf.rs:47]: https://github.com/paritytech/zombienet-sdk/pull/487
Once this is available in zombienet release we can update the code here.

File: polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs (hunk 4 above)
Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs]: done

File: polkadot/zombienet-sdk-tests/tests/utils.rs (hunk 3 above)
Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/utils.rs:67]: Created PR https://github.com/paritytech/zombienet-sdk/pull/488, fyi @pepoviola

File: polkadot/zombienet-sdk-tests/tests/functional/beefy_and_mmr.rs [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+	builder.build().map_err(|e| {
+		let errs = e.into_iter().map(|e| e.to_string()).collect::<Vec<_>>().join(" ");
+		anyhow!("config errs: {errs}")
+	})
+}
+
+/// Verify BEEFY finalized heads across all validator nodes
+async fn verify_beefy_finalized_heads(
+	network: &Network<zombienet_sdk::LocalFileSystem>,
+	node_names: Vec<&str>,
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/beefy_and_mmr.rs]:
```suggestion
	node_names: &[&str],
```
Let's use a slice for args here.

File: polkadot/zombienet-sdk-tests/tests/functional/beefy_and_mmr.rs [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+	}
+
+	Ok(())
+}
+
+/// Generate and verify MMR proofs across all validators
+async fn verify_mmr_proofs(
+	network: &Network<zombienet_sdk::LocalFileSystem>,
+	node_name: &str,
+	validator_names: Vec<&str>,
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/beefy_and_mmr.rs]:
```suggestion
	validator_names: &[&str],
```

File: polkadot/zombienet-sdk-tests/tests/functional/beefy_and_mmr.rs [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+
+	// Get MMR root using RPC
+	let mut root_params = RpcParams::new();
+	root_params.push(format!("{:?}", at_block_hash))?;
+	let root: String = rpc_client.request("mmr_root", root_params).await?;
+	log::info!("MMR root at block 21: {}", root);
+
+	// Generate proof using RPC
+	let mut proof_params = RpcParams::new();
+	proof_params.push(vec![1u32, 9u32, 20u32])?;
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/beefy_and_mmr.rs]:
```suggestion
	proof_params.push(vec![1u32, 9, 20])?;
```
once the first type is defined you can omit the rest.

File: polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+		.await?;
+	assert!(result.success(), "Old node should have no failed recoveries");
+
+	// Check recovery failure metrics
+	new_node
+		.wait_metric_with_timeout(
+			POLKADOT_PARACHAIN_AVAILABILITY_RECOVERY_RECOVERIES_FINISHED_METRIC,
+			|v| v == 0.0,
+			10u64,
+		)
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs]: This metric is a counter, so don't make sense to _wait_ if we want to assert is `0`.

File: polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+		.await
+		.map_err(|e| anyhow!("New node has recovery failures: {}", e))?;
+	old_node
+		.wait_metric_with_timeout(
+			POLKADOT_PARACHAIN_AVAILABILITY_RECOVERY_RECOVERIES_FINISHED_METRIC,
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs]: Same here.

File: polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+			|v| v >= 10.0,
+			400u64,
+		)
+		.await
+		.map_err(|e| anyhow!("Old node fetched chunks too low: {}", e))?;
+
+	new_node
+		.wait_metric_with_timeout(
+			POLKADOT_PARACHAIN_FETCHED_FAILED_CHUNKS_TOTAL_METRIC,
+			|v| v == 0.0,
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs]: Same.

File: polkadot/zombienet-sdk-tests/tests/functional/coretime_shared_core.rs [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+	// Add 4 glutton parachains (2000-2003) without auto-onboarding.
+	for para_id in PARA_IDS {
+		let chain_name = format!("glutton-westend-local-{para_id}");
+		let collator_name = format!("collator-{para_id}");
+		let col_image = col_image.clone();
+
+		builder = builder.with_parachain(|p| {
+			p.with_id(para_id)
+				.onboard_as_parachain(false)
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/coretime_shared_core.rs]:
```suggestion
				.with_registration_strategy(RegistrationStrategy::Manual)
```
I think you want to register manually the paras, set `false`  to _onboard_as_parachain_ will add them to the genesis as parathread.

File: polkadot/zombienet-sdk-tests/tests/functional/dispute_freshly_finalized.rs [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+	let network = initialize_network(config).await?;
+
+	let malus = network.get_node("malus")?;
+	let honest = network.get_node("honest-0")?;
+	let relay_client = honest.wait_client().await?;
+
+	// Check authority status and peers
+	log::info!("Checking node roles");
+	malus
+		.wait_metric_with_timeout("node_roles", |v| v == 4.0, 30u64)
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/dispute_freshly_finalized.rs]: Maybe we can expose this (`node_roles`) as const in the sdk direclty. (for followup pr).

File: polkadot/zombienet-sdk-tests/tests/functional/parachains_max_tranche0.rs [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+	let _ = env_logger::try_init_from_env(
+		env_logger::Env::default().filter_or(env_logger::DEFAULT_FILTER_ENV, "info"),
+	);
+
+	let config = build_network_config()?;
+	let network = initialize_network(config).await?;
+
+	// Check authority status
+	log::info!("Checking validator node roles");
+	for i in [0, 1, 3, 4, 5, 6, 7] {
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/functional/parachains_max_tranche0.rs]:
```suggestion
	for i in 0..NUM_VALIDATORS {
```

File: polkadot/zombienet-sdk-tests/tests/smoke/coretime_smoke.rs [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+				.with_default_image(culumus_image.as_str())
+				.with_chain("coretime-rococo-local")
+				.with_collator(|n| {
+					n.with_name("coretime-collator")
+						.with_args(vec![("-lruntime=debug,xcm=trace").into()])
+				})
+		})
+		.with_parachain(|p| {
+			p.with_id(TEST_PARA_ID)
+				.onboard_as_parachain(false)
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/smoke/coretime_smoke.rs]: In the original test we don't register the para as parathread as will be the result here. In term of core assignment is the same, but I think changing to _registration manually_ will be more clear and follow the same behavior we use in the original test.

File: polkadot/zombienet-sdk-tests/tests/smoke/parachains_upgrade_smoke.rs [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+		.with_relaychain(|r| {
+			r.with_chain("rococo-local")
+				.with_default_command("polkadot")
+				.with_default_image(polkadot_image.as_str())
+				.with_node(|node| node.with_name("alice"))
+				.with_node(|node| node.with_name("bob"))
+		})
+		.with_parachain(|p| {
+			p.with_id(PARA_ID)
+				.cumulus_based(true)
```
Comment (pepoviola) [inline polkadot/zombienet-sdk-tests/tests/smoke/parachains_upgrade_smoke.rs]:
```suggestion
```
this is the default.

**Review by pepoviola — COMMENTED** [review]
Looks good @DenzelPenzel, some inline comments.

File: polkadot/zombienet-sdk-tests/tests/functional/beefy_and_mmr.rs (hunk 6 above)
Comment (mchristou) [reply polkadot/zombienet-sdk-tests/tests/functional/beefy_and_mmr.rs]: done

File: polkadot/zombienet-sdk-tests/tests/functional/parachains_max_tranche0.rs (hunk 13 above)
Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/functional/parachains_max_tranche0.rs]: done

File: polkadot/zombienet-sdk-tests/tests/smoke/parachains_upgrade_smoke.rs (hunk 15 above)
Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/smoke/parachains_upgrade_smoke.rs]: done

File: polkadot/zombienet-sdk-tests/tests/functional/dispute_freshly_finalized.rs (hunk 12 above)
Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/functional/dispute_freshly_finalized.rs]: Yes, for now I’ve created a `const NODE_ROLES_METRIC` in the utils.rs file, but ideally it should be moved to the SDK

File: polkadot/zombienet-sdk-tests/tests/functional/coretime_shared_core.rs (hunk 11 above)
Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/functional/coretime_shared_core.rs]: done!

File: polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs (hunk 8 above)
Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs]: done

File: polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs (hunk 10 above)
Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/functional/chunk_fetching_network_compatibility.rs]: done! change `wait_metric_with_timeout` -> `wait_metric`

File: polkadot/zombienet-sdk-tests/tests/smoke/coretime_smoke.rs (hunk 14 above)
Comment (DenzelPenzel) [reply polkadot/zombienet-sdk-tests/tests/smoke/coretime_smoke.rs]: done

**Comment by pepoviola** [comment]
@DenzelPenzel closing here, thx!
